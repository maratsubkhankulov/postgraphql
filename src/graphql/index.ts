import createGraphQLSchema from './schema/createGQLSchema'
import createGraphQLHTTPRequestHandler from './http/createGraphQLHTTPRequestHandler'

export {
  createGraphQLSchema,
  createGraphQLHTTPRequestHandler,
}
