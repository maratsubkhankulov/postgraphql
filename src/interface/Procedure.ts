import Paginator from './Paginator'
import Type from './type/Type'
import ObjectType from './type/ObjectType'

/**
 * A procedure is any remote function which our interface knows about.
 * Procedures allow users to execute arbitrary code from their APIs.
 */
// TODO: Rethink the procedure interface from a non-Postgres perspective.
// Having paginator output for mutations can be wierd. Maybe paginators
// should be instance based?
// TODO: Consider splitting up a `Procedure` into multiple types. One for
// mutations, one for queries/views.
interface Procedure {
  /**
   * The required name of our procedure.
   */
  readonly name: string

  /**
   * The optional description of our procedure.
   */
  readonly description?: string | undefined

  /**
   * Indicates whether or not the procedure mutates data.
   *
   * The reason this is called `isStable` vs. `isMutation` is becuase we want
   * to assume the procedure is a mutation if the property is not defined.
   */
  readonly isStable?: boolean

  /**
   * The input type we will use for our procedure. We use an object type
   * because that forces the procedure to have multiple arguments, *and* named
   * arguments. When building APIs, named arguments are much better than
   * ordered.
   */
  readonly inputType: ObjectType

  /**
   * The output `Procedure` strategy. A procedure could output one of two
   * things. A single value, or a set of values. Expressed by a paginator.
   */
  // TODO: Does an unstable procedure with a paginator output make sense?
  // Perhaps only stable procedures should have paginators.
  readonly output: Procedure.SingleOutput<mixed> | Procedure.PaginatorOutput<mixed>
}

namespace Procedure {
  /**
   * The output object for a procedure that returns a single value.
   */
  export interface SingleOutput<TOutputValue> {
    /**
     * A constant tag to identify this output type from other output types.
     */
    readonly kind: 'SINGLE'

    /**
     * The type of the value the paginator will output.
     */
    readonly outputType: Type<TOutputValue>

    /**
     * Actually execute our procedure.
     */
    execute (context: mixed, input: ObjectType.Value): Promise<TOutputValue>
  }

  /**
   * The output object for a procedure that returns a set of values.
   */
  export interface PaginatorOutput<TOutputItemValue> {
    /**
     * The tag necessary for distinguishing this output type from other output
     * types.
     */
    readonly kind: 'PAGINATOR'

    /**
     * The paginator to be used in executing our procedure. Takes the value
     * specified by `Procedure#inputType` as the input.
     */
    readonly paginator: Paginator<ObjectType.Value, TOutputItemValue>
  }
}

export default Procedure
