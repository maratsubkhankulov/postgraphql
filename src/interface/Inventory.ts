import Context from './Context'
import NamedType from './type/NamedType'
import AliasType from './type/AliasType'
import ObjectType from './type/ObjectType'
import Collection from './collection/Collection'
import Relation from './collection/Relation'
import Procedure from './Procedure'

type ContextAssignmentFn = (context: Context) => void | Promise<void>

/**
 * In order to build awesome tools from any database we need an abstract
 * static data definition language. Inventory is the root level interface
 * to this definition.
 *
 * If a database is strongly typed (like PostgreSQL) it may implement the
 * inventory interfaces to expose its underlying data.
 *
 * Believe it or not, a `Inventory` and all of the objects inside it are
 * *mutable*. Scary, I know.
 */
class Inventory {
  private _contextAssignments: Array<ContextAssignmentFn> = []
  private _types = new Map<string, NamedType<mixed>>()
  private _collections = new Map<string, Collection>()
  private _relations = new Map<string, Relation<mixed>>()
  private _procedures = new Map<string, Procedure>()

  /**
   * Adds a function which will assign some values to a context object when it
   * gets created.
   */
  // TODO: Clean this API up! The current context API is terrible.
  public addContextAssignment (contextAssignment: ContextAssignmentFn): this {
    this._contextAssignments.push(contextAssignment)
    return this
  }

  /**
   * Creates a context object by running all of the context assignments in
   * parallel. A context will be valid for one single request “session.”
   */
  public async createContext (): Promise<Context> {
    const context = new Context()
    await Promise.all(this._contextAssignments.map(contextAssignment => contextAssignment(context)))
    return context
  }

  /**
   * Adds a single collection to our inventory. If a collection with the same
   * name already exists, an error is thrown. If the collection has a
   * different inventory, an error is thrown.
   *
   * We will also add the type for this collection to the inventory.
   */
  public addCollection (collection: Collection): this {
    const { name } = collection

    if (this._collections.has(name))
      throw new Error(`Collection of name '${name}' already exists in the inventory.`)

    this._collections.set(name, collection)

    return this
  }

  /**
   * Gets all of the collections in our system. A collection will always have
   * an object type, which should be accessible through `getTypes`.
   *
   * Two collections in this set should not have the same name.
   */
  public getCollections (): Array<Collection> {
    return Array.from(this._collections.values())
  }

  /**
   * Gets a single collection by name.
   */
  public getCollection (name: string): Collection | undefined {
    return this._collections.get(name)
  }

  /**
   * Determines if a *specific* collection has been added to the inventory. If
   * the exact reference to the collection argument exists in the inventory this
   * method returns true, otherwise it returns false.
   */
  public hasCollection (collection: Collection): boolean {
    return this._collections.get(collection.name) === collection
  }

  /**
   * Adds a single relation to our inventory. If the related collections are not
   * members of this inventory we fail with an error.
   */
  public addRelation (relation: Relation<mixed>): this {
    const { name, tailCollection, headCollectionKey } = relation

    if (!this.hasCollection(tailCollection))
      throw new Error(`Tail collection named '${tailCollection.name}' is not in this inventory.`)

    if (!this.hasCollection(headCollectionKey.collection))
      throw new Error(`Head collection named '${headCollectionKey.collection.name}' is not in this inventory.`)

    if (this._relations.has(name))
      throw new Error(`Relation of name '${name}' already exists in the inventory.`)

    this._relations.set(name, relation)

    return this
  }

  /**
   * Gets all of the relations in our inventory. A relationship is formed when
   * the values of one collection reference another. A relation can be used to
   * track that reference and perform operations using that reference.
   *
   * In a graph representation of our inventory, collections would be nodes and
   * relations would be directed edges.
   */
  public getRelations (): Array<Relation<mixed>> {
    return Array.from(this._relations.values())
  }

  /**
   * Adds a single `Procedure` to our inventory. If a procedure of the same
   * name already exists, an error will be thrown.
   */
  public addProcedure (procedure: Procedure): this {
    const { name } = procedure

    if (this._procedures.has(name))
      throw new Error(`Procedure of name '${name}' already exists in the inventory.'`)

    this._procedures.set(name, procedure)

    return this
  }

  /**
   * Returns all of the procedures in our inventory.
   */
  public getProcedures (): Array<Procedure> {
    return Array.from(this._procedures.values())
  }
}

export default Inventory
