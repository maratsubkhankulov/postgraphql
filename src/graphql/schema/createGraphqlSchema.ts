import { GraphQLSchema, GraphQLObjectType, GraphQLFieldConfig } from 'graphql'
import { Inventory } from '../../interface'
import buildObject from '../utils/buildObject'
import createNodeFieldEntry from './node/createNodeFieldEntry'
import getCollectionType from './collection/getCollectionType'
import createCollectionQueryFieldEntries from './collection/createCollectionQueryFieldEntries'
import BuildToken from './BuildToken'

export type SchemaOptions = {
  // The exact name for the node id field. In the past Relay wanted this to
  // be `id`, but there are some movements in the GraphQL standard to
  // standardize an `__id` field.
  nodeIdFieldName?: string,
}

/**
 * Creates a GraphQL schema using an instance of `Inventory`.
 */
export default function createGraphqlSchema (inventory: Inventory, options: SchemaOptions = {}): GraphQLSchema {
  // We take our user-friendly arguments to `createGraphqlSchema` and convert them
  // into a build token. One nice side effect of always creating our own
  // build token object is that we have the guarantee that every build token
  // will always maintain its own memoization map.
  const buildToken: BuildToken = {
    inventory,
    options: {
      // The default node id field name is `__id` as it is the emerging
      // standard.
      nodeIdFieldName: options.nodeIdFieldName || '__id',
    },
  }

  return new GraphQLSchema({
    query: createQueryType(buildToken),
    types: [
      // Make sure to always include the types for our collections, even if
      // they have no other output.
      ...inventory.getCollections().map(collection => getCollectionType(buildToken, collection)),
    ],
  })
}

// TODO: doc
function createQueryType <T>(buildToken: BuildToken): GraphQLObjectType<T> {
  const { inventory } = buildToken
  return new GraphQLObjectType({
    name: 'Query',
    // TODO: description
    fields: buildObject<GraphQLFieldConfig<T, mixed>>(
      [
        createNodeFieldEntry(buildToken),
      ],
      inventory
        .getCollections()
        .map(collection => createCollectionQueryFieldEntries(buildToken, collection))
        .reduce((a, b) => a.concat(b), []),
    ),
  })
}
