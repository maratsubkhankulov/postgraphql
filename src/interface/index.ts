import Context from './Context'
import Inventory from './Inventory'
import Collection from './collection/Collection'
import CollectionKey from './collection/CollectionKey'
import Paginator from './collection/Paginator'
import Condition, { conditionHelpers } from './collection/Condition'
import Relation from './collection/Relation'
import Type from './type/Type'
import NullableType from './type/NullableType'
import ListType from './type/ListType'
import NamedType, { isNamedType } from './type/NamedType'
import AliasType from './type/AliasType'
import EnumType from './type/EnumType'
import ObjectType from './type/ObjectType'
import booleanType from './type/primitive/booleanType'
import integerType from './type/primitive/integerType'
import floatType from './type/primitive/floatType'
import stringType from './type/primitive/stringType'

export {
  Context,
  Inventory,
  Collection,
  CollectionKey,
  Paginator,
  Condition,
  conditionHelpers,
  Relation,
  Type,
  NullableType,
  ListType,
  NamedType,
  isNamedType,
  AliasType,
  EnumType,
  ObjectType,
  booleanType,
  integerType,
  floatType,
  stringType,
}
