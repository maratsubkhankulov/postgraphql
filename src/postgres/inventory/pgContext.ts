// TODO: Refactor this module, it has code smellâ€¦

import createDebugger = require('debug')
import { PoolConfig, Pool, Client } from 'pg'
import { Context } from '../../interface'

const debugQuery = createDebugger('postgraphql:postgres:query')

const $$pgClient = Symbol('postgres/client')

/**
 * Creates a function that will assign a Postgres client to a context variable.
 * The config passed in will be used to create a pool which will connect
 * clients.
 */
export function createPGContextAssignment (pool: Pool) {
  return async (context: Context) => {
    const client = await pool.connect()

    // If the client query debugger is enabled, log all of our Postgres
    // queries.
    if (debugQuery.enabled) {
      const originalQuery = client.query
      client.query = function debuggingQuery (...args: Array<any>) {
        // If possible we never want to log the variables. Only the query text.
        // If we logged variables we may log a password in plaintext.
        if (typeof args[0] === 'string') debugQuery(args[0])
        else if (typeof args[0] === 'object') debugQuery(args[0].text)

        return originalQuery.apply(this, args)
      }
    }

    context.addCleanupFunction(() => client.release())
    context.addDependency($$pgClient, client)
  }
}

/**
 * Creates a new context and adds the Postgres client. This function is mainly
 * used for testing.
 */
export function createPGContext (client: Client): Context {
  const context = new Context()
  context.addDependency($$pgClient, client)
  return context
}

/**
 * Retrieves a Postgres client from a context, throwing an error if such a
 * client does not exist.
 */
export function pgClientFromContext (context: Context): Client {
  const client = context.getDependency($$pgClient)

  if (client == null || !(client instanceof Client))
    throw new Error('Must pass add a Postgres client as a dependency of the context.')

  return client
}
