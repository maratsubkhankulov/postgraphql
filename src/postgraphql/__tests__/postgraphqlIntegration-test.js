import { resolve as resolvePath } from 'path'
import { readFile, readdirSync } from 'fs'
import { GraphQLSchema, printSchema, graphql } from 'graphql'
import { Inventory, Context } from '../../interface'
import withPGClient from '../../postgres/__tests__/fixtures/withPGClient'
import { introspectDatabase } from '../../postgres/introspection'
import addPGCatalogToInventory from '../../postgres/inventory/addPGCatalogToInventory'
import { createPGContext } from '../../postgres/inventory/pgContext'
import createGraphqlSchema from '../../graphql/schema/createGraphqlSchema'

const kitchenSinkData = new Promise((resolve, reject) => {
  readFile(resolvePath(__dirname, '../../../resources/kitchen-sink-data.sql'), (error, data) => {
    if (error) reject(error)
    else resolve(data.toString())
  })
})

/**
 * @type {GraphQLSchema}
 */
let schema1, schema2

beforeAll(withPGClient(async client => {
  try {
    const catalog = await introspectDatabase(client, ['a', 'b', 'c'])

    const inventory1 = new Inventory()
    addPGCatalogToInventory(inventory1, catalog)
    schema1 = createGraphqlSchema(inventory1)

    const inventory2 = new Inventory()
    addPGCatalogToInventory(inventory2, catalog, { renameIdToRowId: true })
    schema2 = createGraphqlSchema(inventory2, { nodeIdFieldName: 'id' })
  }
  catch (error) {
    console.error(error.stack)
    throw error
  }
}))

test('schema', async () => {
  expect(printSchema(schema1)).toMatchSnapshot()
  expect(printSchema(schema2)).toMatchSnapshot()
})

const queriesDir = resolvePath(__dirname, 'fixtures/queries')

for (const file of readdirSync(queriesDir)) {
  test(`query ${file}`, withPGClient(async client => {
    const query = await new Promise((resolve, reject) => {
      readFile(resolvePath(queriesDir, file), (error, data) => {
        if (error) reject(error)
        else resolve(data.toString())
      })
    })

    await client.query(await kitchenSinkData)
    const context = createPGContext(client)

    const result = await graphql(schema1, query, null, context)

    // Log the errors in our result.
    if (result.errors)
      for (const e of result.errors)
        console.error(e.stack || e)

    expect(result).toMatchSnapshot()
  }))
}
