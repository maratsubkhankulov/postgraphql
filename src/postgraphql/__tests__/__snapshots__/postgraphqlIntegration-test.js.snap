exports[`test query node.graphql 1`] = `
Object {
  "data": Object {
    "a": Object {
      "nodes": Array [
        Object {
          "__id": "WyJwZW9wbGUiLDFd",
          "__typename": "Person",
          "id": 1,
          "name": "John Smith"
        },
        Object {
          "__id": "WyJwZW9wbGUiLDJd",
          "__typename": "Person",
          "id": 2,
          "name": "Sara Smith"
        },
        Object {
          "__id": "WyJwZW9wbGUiLDNd",
          "__typename": "Person",
          "id": 3,
          "name": "Budd Deey"
        },
        Object {
          "__id": "WyJwZW9wbGUiLDRd",
          "__typename": "Person",
          "id": 4,
          "name": "Kathryn Ramirez"
        },
        Object {
          "__id": "WyJwZW9wbGUiLDVd",
          "__typename": "Person",
          "id": 5,
          "name": "Johnny Tucker"
        }
      ]
    },
    "b": Object {
      "nodes": Array [
        Object {
          "__id": "WyJjb21wb3VuZF9rZXlzIiwxLDJd",
          "__typename": "CompoundKey",
          "personId1": 1,
          "personId2": 2
        },
        Object {
          "__id": "WyJjb21wb3VuZF9rZXlzIiwyLDFd",
          "__typename": "CompoundKey",
          "personId1": 2,
          "personId2": 1
        },
        Object {
          "__id": "WyJjb21wb3VuZF9rZXlzIiwyLDNd",
          "__typename": "CompoundKey",
          "personId1": 2,
          "personId2": 3
        },
        Object {
          "__id": "WyJjb21wb3VuZF9rZXlzIiwyLDVd",
          "__typename": "CompoundKey",
          "personId1": 2,
          "personId2": 5
        },
        Object {
          "__id": "WyJjb21wb3VuZF9rZXlzIiw0LDNd",
          "__typename": "CompoundKey",
          "personId1": 4,
          "personId2": 3
        },
        Object {
          "__id": "WyJjb21wb3VuZF9rZXlzIiw0LDRd",
          "__typename": "CompoundKey",
          "personId1": 4,
          "personId2": 4
        }
      ]
    },
    "c": Object {
      "__id": "WyJjb21wb3VuZF9rZXlzIiwyLDNd",
      "__typename": "CompoundKey",
      "personId1": 2,
      "personId2": 3
    },
    "d": Object {
      "__id": "WyJwZW9wbGUiLDJd",
      "__typename": "Person",
      "id": 2,
      "name": "Sara Smith"
    },
    "e": Object {
      "__id": "WyJjb21wb3VuZF9rZXlzIiwxLDJd",
      "__typename": "CompoundKey",
      "personId1": 1,
      "personId2": 2
    },
    "f": Object {
      "__id": "WyJwZW9wbGUiLDVd",
      "__typename": "Person",
      "id": 5,
      "name": "Johnny Tucker"
    },
    "g": null,
    "h": null,
    "i": Object {
      "__id": "WyJwZW9wbGUiLDJd",
      "__typename": "Person",
      "id": 2,
      "name": "Sara Smith"
    },
    "j": Object {
      "__id": "WyJwZW9wbGUiLDVd",
      "__typename": "Person",
      "id": 5,
      "name": "Johnny Tucker"
    },
    "k": null,
    "l": Object {
      "__id": "WyJjb21wb3VuZF9rZXlzIiwyLDNd",
      "__typename": "CompoundKey",
      "personId1": 2,
      "personId2": 3
    },
    "m": Object {
      "__id": "WyJjb21wb3VuZF9rZXlzIiwxLDJd",
      "__typename": "CompoundKey",
      "personId1": 1,
      "personId2": 2
    },
    "n": null
  }
}
`;

exports[`test query relation-tail-head.graphql 1`] = `
Object {
  "data": Object {
    "allCompoundKeys": Object {
      "nodes": Array [
        Object {
          "extra": null,
          "personByPersonId1": Object {
            "createdAt": null,
            "email": "john.smith@email.com",
            "name": "John Smith"
          },
          "personByPersonId2": Object {
            "createdAt": null,
            "email": "sara.smith@email.com",
            "name": "Sara Smith"
          },
          "personId1": 1,
          "personId2": 2
        },
        Object {
          "extra": false,
          "personByPersonId1": Object {
            "createdAt": null,
            "email": "sara.smith@email.com",
            "name": "Sara Smith"
          },
          "personByPersonId2": Object {
            "createdAt": null,
            "email": "john.smith@email.com",
            "name": "John Smith"
          },
          "personId1": 2,
          "personId2": 1
        },
        Object {
          "extra": null,
          "personByPersonId1": Object {
            "createdAt": null,
            "email": "sara.smith@email.com",
            "name": "Sara Smith"
          },
          "personByPersonId2": Object {
            "createdAt": null,
            "email": "budd.deey@email.com",
            "name": "Budd Deey"
          },
          "personId1": 2,
          "personId2": 3
        },
        Object {
          "extra": true,
          "personByPersonId1": Object {
            "createdAt": null,
            "email": "sara.smith@email.com",
            "name": "Sara Smith"
          },
          "personByPersonId2": Object {
            "createdAt": null,
            "email": "johnny.tucker@email.com",
            "name": "Johnny Tucker"
          },
          "personId1": 2,
          "personId2": 5
        },
        Object {
          "extra": true,
          "personByPersonId1": Object {
            "createdAt": null,
            "email": "kathryn.ramirez@email.com",
            "name": "Kathryn Ramirez"
          },
          "personByPersonId2": Object {
            "createdAt": null,
            "email": "budd.deey@email.com",
            "name": "Budd Deey"
          },
          "personId1": 4,
          "personId2": 3
        },
        Object {
          "extra": false,
          "personByPersonId1": Object {
            "createdAt": null,
            "email": "kathryn.ramirez@email.com",
            "name": "Kathryn Ramirez"
          },
          "personByPersonId2": Object {
            "createdAt": null,
            "email": "kathryn.ramirez@email.com",
            "name": "Kathryn Ramirez"
          },
          "personId1": 4,
          "personId2": 4
        }
      ]
    }
  }
}
`;

exports[`test query types.graphql 1`] = `
Object {
  "data": Object {
    "allTypes": Object {
      "nodes": Array [
        Object {
          "bigint": 200,
          "boolean": true,
          "compoundType": Object {
            "a": 1,
            "b": "2",
            "c": "BLUE",
            "d": "4be8a712-3ff7-432e-aa34-fdb43fbd838d",
            "fooBar": 8
          },
          "domain": 5,
          "domain2": 6,
          "enum": "GREEN",
          "nestedCompoundType": Object {
            "a": Object {
              "a": 3,
              "b": "456",
              "c": "RED",
              "d": "aed18400-2a92-46df-8204-b70c728b3520",
              "fooBar": 93
            },
            "b": Object {
              "a": 42,
              "b": "Hello, world!",
              "c": "BLUE",
              "d": "d34df5e0-83f1-11e6-8dd0-abee917ffd1e",
              "fooBar": -5
            },
            "bazBuz": 7
          }
        }
      ]
    }
  }
}
`;

exports[`test schema 1`] = `
"scalar AnInt

scalar AnotherInt

enum Color {
  RED
  GREEN
  BLUE
}

type CompoundKey implements Node {
  __id: ID!
  personId2: Int!
  personId1: Int!
  extra: Boolean
  personByPersonId1: Person
  personByPersonId2: Person
}

type CompoundKeysConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [CompoundKeysEdge]
  nodes: [CompoundKey!]
}

type CompoundKeysEdge {
  cursor: Cursor!
  node: CompoundKey!
}

enum CompoundKeysOrderBy {
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
  NATURAL
  PERSON_ID_2_ASC
  PERSON_ID_2_DESC
  PERSON_ID_1_ASC
  PERSON_ID_1_DESC
  EXTRA_ASC
  EXTRA_DESC
}

# Awesome feature!
type CompoundType {
  a: Int
  b: String
  c: Color
  d: Uuid
  fooBar: Int
}

scalar Cursor

scalar Email

type NestedCompoundType {
  a: CompoundType
  b: CompoundType
  bazBuz: Int
}

interface Node {
  __id: ID!
}

type NonUpdatableView {
  column: Int
}

type NonUpdatableViewsConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [NonUpdatableViewsEdge]
  nodes: [NonUpdatableView!]
}

type NonUpdatableViewsEdge {
  cursor: Cursor!
  node: NonUpdatableView!
}

enum NonUpdatableViewsOrderBy {
  NATURAL
  COLUMN_ASC
  COLUMN_DESC
}

type PageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: Cursor
  endCursor: Cursor
}

type PeopleConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [PeopleEdge]
  nodes: [Person!]
}

type PeopleEdge {
  cursor: Cursor!
  node: Person!
}

enum PeopleOrderBy {
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
  NATURAL
  ID_ASC
  ID_DESC
  NAME_ASC
  NAME_DESC
  ABOUT_ASC
  ABOUT_DESC
  EMAIL_ASC
  EMAIL_DESC
  CREATED_AT_ASC
  CREATED_AT_DESC
}

# Person test comment
type Person implements Node {
  __id: ID!
  id: Int!

  # The personâ€™s name
  name: String!
  about: String
  email: Email
  createdAt: String
}

type Post implements Node {
  __id: ID!
  id: Int!
  headline: String!
  body: String
  authorId: Int
  personByAuthorId: Person
}

type PostsConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [PostsEdge]
  nodes: [Post!]
}

type PostsEdge {
  cursor: Cursor!
  node: Post!
}

enum PostsOrderBy {
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
  NATURAL
  ID_ASC
  ID_DESC
  HEADLINE_ASC
  HEADLINE_DESC
  BODY_ASC
  BODY_DESC
  AUTHOR_ID_ASC
  AUTHOR_ID_DESC
}

type Query {
  node(__id: ID!): Node
  allNonUpdatableViews(orderBy: NonUpdatableViewsOrderBy = NATURAL, before: Cursor, after: Cursor, first: Int, last: Int): NonUpdatableViewsConnection
  allPosts(orderBy: PostsOrderBy = PRIMARY_KEY_ASC, before: Cursor, after: Cursor, first: Int, last: Int): PostsConnection
  post(__id: ID!): Post
  postById(id: Int!): Post
  allTypes(orderBy: TypesOrderBy = PRIMARY_KEY_ASC, before: Cursor, after: Cursor, first: Int, last: Int): TypesConnection
  types(__id: ID!): Types
  typesById(id: Int!): Types
  allUpdatableViews(orderBy: UpdatableViewsOrderBy = NATURAL, before: Cursor, after: Cursor, first: Int, last: Int): UpdatableViewsConnection
  allCompoundKeys(orderBy: CompoundKeysOrderBy = PRIMARY_KEY_ASC, before: Cursor, after: Cursor, first: Int, last: Int): CompoundKeysConnection
  compoundKey(__id: ID!): CompoundKey
  compoundKeyByPersonId1AndPersonId2(personId1: Int!, personId2: Int!): CompoundKey
  allPeople(orderBy: PeopleOrderBy = PRIMARY_KEY_ASC, before: Cursor, after: Cursor, first: Int, last: Int): PeopleConnection
  person(__id: ID!): Person
  personByEmail(email: Email): Person
  personById(id: Int!): Person
}

type Types implements Node {
  __id: ID!
  id: Int!
  bigint: Int
  boolean: Boolean
  varchar: String
  enum: Color
  domain: AnInt
  domain2: AnotherInt
  compoundType: CompoundType
  nestedCompoundType: NestedCompoundType
}

type TypesConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [TypesEdge]
  nodes: [Types!]
}

type TypesEdge {
  cursor: Cursor!
  node: Types!
}

enum TypesOrderBy {
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
  NATURAL
  ID_ASC
  ID_DESC
  BIGINT_ASC
  BIGINT_DESC
  BOOLEAN_ASC
  BOOLEAN_DESC
  VARCHAR_ASC
  VARCHAR_DESC
  ENUM_ASC
  ENUM_DESC
  DOMAIN_ASC
  DOMAIN_DESC
  DOMAIN2_ASC
  DOMAIN2_DESC
  COMPOUND_TYPE_ASC
  COMPOUND_TYPE_DESC
  NESTED_COMPOUND_TYPE_ASC
  NESTED_COMPOUND_TYPE_DESC
}

# YOYOYO!!
type UpdatableView {
  x: Int
  name: String
  description: String

  # This is constantly 2
  constant: Int
}

type UpdatableViewsConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [UpdatableViewsEdge]
  nodes: [UpdatableView!]
}

type UpdatableViewsEdge {
  cursor: Cursor!
  node: UpdatableView!
}

enum UpdatableViewsOrderBy {
  NATURAL
  X_ASC
  X_DESC
  NAME_ASC
  NAME_DESC
  DESCRIPTION_ASC
  DESCRIPTION_DESC
  CONSTANT_ASC
  CONSTANT_DESC
}

# A universally unique identifier as defined by [RFC 4122](https://tools.ietf.org/html/rfc4122).
scalar Uuid
"
`;

exports[`test schema 2`] = `
"scalar AnInt

scalar AnotherInt

enum Color {
  RED
  GREEN
  BLUE
}

type CompoundKey implements Node {
  id: ID!
  personId2: Int!
  personId1: Int!
  extra: Boolean
  personByPersonId1: Person
  personByPersonId2: Person
}

type CompoundKeysConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [CompoundKeysEdge]
  nodes: [CompoundKey!]
}

type CompoundKeysEdge {
  cursor: Cursor!
  node: CompoundKey!
}

enum CompoundKeysOrderBy {
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
  NATURAL
  PERSON_ID_2_ASC
  PERSON_ID_2_DESC
  PERSON_ID_1_ASC
  PERSON_ID_1_DESC
  EXTRA_ASC
  EXTRA_DESC
}

# Awesome feature!
type CompoundType {
  a: Int
  b: String
  c: Color
  d: Uuid
  fooBar: Int
}

scalar Cursor

scalar Email

type NestedCompoundType {
  a: CompoundType
  b: CompoundType
  bazBuz: Int
}

interface Node {
  id: ID!
}

type NonUpdatableView {
  column: Int
}

type NonUpdatableViewsConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [NonUpdatableViewsEdge]
  nodes: [NonUpdatableView!]
}

type NonUpdatableViewsEdge {
  cursor: Cursor!
  node: NonUpdatableView!
}

enum NonUpdatableViewsOrderBy {
  NATURAL
  COLUMN_ASC
  COLUMN_DESC
}

type PageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: Cursor
  endCursor: Cursor
}

type PeopleConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [PeopleEdge]
  nodes: [Person!]
}

type PeopleEdge {
  cursor: Cursor!
  node: Person!
}

enum PeopleOrderBy {
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
  NATURAL
  ID_ASC
  ID_DESC
  NAME_ASC
  NAME_DESC
  ABOUT_ASC
  ABOUT_DESC
  EMAIL_ASC
  EMAIL_DESC
  CREATED_AT_ASC
  CREATED_AT_DESC
}

# Person test comment
type Person implements Node {
  id: ID!
  rowId: Int!

  # The personâ€™s name
  name: String!
  about: String
  email: Email
  createdAt: String
}

type Post implements Node {
  id: ID!
  rowId: Int!
  headline: String!
  body: String
  authorId: Int
  personByAuthorId: Person
}

type PostsConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [PostsEdge]
  nodes: [Post!]
}

type PostsEdge {
  cursor: Cursor!
  node: Post!
}

enum PostsOrderBy {
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
  NATURAL
  ID_ASC
  ID_DESC
  HEADLINE_ASC
  HEADLINE_DESC
  BODY_ASC
  BODY_DESC
  AUTHOR_ID_ASC
  AUTHOR_ID_DESC
}

type Query {
  node(id: ID!): Node
  allNonUpdatableViews(orderBy: NonUpdatableViewsOrderBy = NATURAL, before: Cursor, after: Cursor, first: Int, last: Int): NonUpdatableViewsConnection
  allPosts(orderBy: PostsOrderBy = PRIMARY_KEY_ASC, before: Cursor, after: Cursor, first: Int, last: Int): PostsConnection
  post(id: ID!): Post
  postByRowId(rowId: Int!): Post
  allTypes(orderBy: TypesOrderBy = PRIMARY_KEY_ASC, before: Cursor, after: Cursor, first: Int, last: Int): TypesConnection
  types(id: ID!): Types
  typesByRowId(rowId: Int!): Types
  allUpdatableViews(orderBy: UpdatableViewsOrderBy = NATURAL, before: Cursor, after: Cursor, first: Int, last: Int): UpdatableViewsConnection
  allCompoundKeys(orderBy: CompoundKeysOrderBy = PRIMARY_KEY_ASC, before: Cursor, after: Cursor, first: Int, last: Int): CompoundKeysConnection
  compoundKey(id: ID!): CompoundKey
  compoundKeyByPersonId1AndPersonId2(personId1: Int!, personId2: Int!): CompoundKey
  allPeople(orderBy: PeopleOrderBy = PRIMARY_KEY_ASC, before: Cursor, after: Cursor, first: Int, last: Int): PeopleConnection
  person(id: ID!): Person
  personByEmail(email: Email): Person
  personByRowId(rowId: Int!): Person
}

type Types implements Node {
  id: ID!
  rowId: Int!
  bigint: Int
  boolean: Boolean
  varchar: String
  enum: Color
  domain: AnInt
  domain2: AnotherInt
  compoundType: CompoundType
  nestedCompoundType: NestedCompoundType
}

type TypesConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [TypesEdge]
  nodes: [Types!]
}

type TypesEdge {
  cursor: Cursor!
  node: Types!
}

enum TypesOrderBy {
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
  NATURAL
  ID_ASC
  ID_DESC
  BIGINT_ASC
  BIGINT_DESC
  BOOLEAN_ASC
  BOOLEAN_DESC
  VARCHAR_ASC
  VARCHAR_DESC
  ENUM_ASC
  ENUM_DESC
  DOMAIN_ASC
  DOMAIN_DESC
  DOMAIN2_ASC
  DOMAIN2_DESC
  COMPOUND_TYPE_ASC
  COMPOUND_TYPE_DESC
  NESTED_COMPOUND_TYPE_ASC
  NESTED_COMPOUND_TYPE_DESC
}

# YOYOYO!!
type UpdatableView {
  x: Int
  name: String
  description: String

  # This is constantly 2
  constant: Int
}

type UpdatableViewsConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [UpdatableViewsEdge]
  nodes: [UpdatableView!]
}

type UpdatableViewsEdge {
  cursor: Cursor!
  node: UpdatableView!
}

enum UpdatableViewsOrderBy {
  NATURAL
  X_ASC
  X_DESC
  NAME_ASC
  NAME_DESC
  DESCRIPTION_ASC
  DESCRIPTION_DESC
  CONSTANT_ASC
  CONSTANT_DESC
}

# A universally unique identifier as defined by [RFC 4122](https://tools.ietf.org/html/rfc4122).
scalar Uuid
"
`;
